File: cpp/day_22a.cpp
#include <algorithm>
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <vector>
#include <deque>
#include <functional>

int main() {
  std::fstream file{"../input/day_22_input"};
  std::string line;
  std::vector<std::deque<int>> player_decks;
  std::vector<int> players;
  while(std::getline(file, line)) {
    line.erase(std::remove_if(std::begin(line), std::end(line), [](const char c) {return !isprint(c);}), std::end(line));
    if(line == "") continue;
    if(line.substr(0, 6) == "Player") {
      player_decks.emplace_back();
      players.push_back(std::stoi(line.substr(7, line.size() - 8)));
    } else {
      player_decks.back().push_back(std::stoi(line));
    }
  }
  int total_players = players.size();

  // Play
  while (player_decks.size() > 1) {
    size_t winning_player = -1;
    int hightest_card = -1;
    std::vector<int> top_cards(player_decks.size());
    for(int i = 0; i < player_decks.size(); i++) {
      if(hightest_card < player_decks[i].front() ) {
        hightest_card = player_decks[i].front();
        winning_player = i;
      }
      top_cards[i] = player_decks[i].front();
      player_decks[i].pop_front();
    }
    std::sort(std::begin(top_cards), std::end(top_cards), std::greater<int>());
    for(const auto& top_card : top_cards) {
      player_decks[winning_player].push_back(top_card);
    }
    for(int i = 0; i < player_decks.size(); i++) {
      if(player_decks[i].empty()) {
        player_decks.erase(std::begin(player_decks) + i);
        players.erase(std::begin(players) + i);
        i--;
      }
    }
  }

  long long score = 0;
  for (const auto& player_deck : player_decks) {
    for(int i = 0; i < player_deck.size(); i++) {
      score += ((i+1) * player_deck[player_deck.size() - 1 - i]);
    }
  }
  std::cout << score << '\n';
  return score;
}

File: python/day_22a.py
#!/usr/bin/python3


def main():
    file = open("../input/day_22_input")
    players = list()
    player_decks = list()
    for line in file:
        line = line.strip("\n")
        if line == "":
            continue
        if line[0:6] == "Player":
            player_decks.append([])
            players.append(int(line[7 : len(line) - 1]))
        else:
            player_decks[-1].append(int(line))
    total_players = len(players)

    while len(player_decks) > 1:
        winning_player = -1
        hightest_card = -1
        top_cards = list()
        for i in range(0, len(player_decks)):
            if hightest_card < player_decks[i][0]:
                hightest_card = player_decks[i][0]
                winning_player = i
            top_cards.append(player_decks[i][0])
            player_decks[i].pop(0)
        top_cards.sort(reverse=True)
        for top_card in top_cards:
            player_decks[winning_player].append(top_card)
        for i in range(0, len(player_decks)):
            if len(player_decks[i]) == 0:
                player_decks.pop(i)
                players.pop(i)
                i = i - 1
    score = 0
    for player_deck in player_decks:
        for i in range(0, len(player_deck)):
            score = score + ((i + 1) * player_deck[len(player_deck) - 1 - i])
    print(score)
    return score


if __name__ == "__main__":
    main()
